Rewrite English man pages in mdoc, update for Nvim (old)

Pyrohh opened the pull request:
refs #1826

I used this guide, which proved to be very helpful: http://manpages.bsd.lv/
It's very comprehensive, but mdoc is quite simple so it only took a few hours.

The man pages can be previewed locally using `man -l`, but note that some options are missing currently.

`xxd.1` hasn't been touched for now as I don't think it's going to be included, as mentioned <br>by @ZyX-I https://github.com/neovim/neovim/issues/1646#issuecomment-73593266.

**Edit**: updated post due to rewrite in mdoc.

splinterofchaos commented on runtime/doc/manpages/nvim.1.md:
```diff
@@ -0,0 +1,261 @@
+`+/`<*pat*>
+  For the first file, the cursor will be positioned on the first
+  occurrence of *pat*.
+  See ":help search-pattern" for the available search patterns.
```
I think both `+[num] and `+/<pat>` could be more accurately described as examples of `-c`,

Pyrohh replied:
I think there's a stray backtic in your comment, but besides that I'm not clear on what you're suggesting.

ZyX-I replied:
Try `:[num]` in NeoVim. It will result in you going to line number `[num]`. I cannot say where this behaviour is described, but entering command-line range *without* entering a command will result in a jump to the last line in the range.

Thus `+1` and `+/pat` are the same thing as `+cmd`, not something special. `+` *without* arguments *is* a special case: empty command does not normally act as `+$` if you e.g. do `:<CR>` or `:execute ''<CR>`.

Pyrohh replied:
I see your point, I'll update the man pages when I can.

splinterofchaos commented on runtime/doc/manpages/nvim.1.md:
```diff
@@ -0,0 +1,261 @@
+`SHELL`
+  Used to set the 'shell' option, which determines the shell used by the
+  ":terminal" command.
+
```
`$NVIM_TUI_CURSOR_SHAPE`

Pyrohh replied:
:+1: I'll also mention the TRUE_COLOR one.

justinmk commented on runtime/doc/Makefile:
```diff
@@ -247,30 +247,30 @@ HTMLS = \
 MANPAGES = \
-	manpages/vim.man \
-	manpages/vimtutor.man \
+	manpages/nvim.man \
```
I'm excited that we have a man page now :)

Pyrohh replied:
We're *finally* a __real__ project.

aktau commented:
I usually use the asciidoc and xmlto combo to convert Markdown documents into manpages (and HTML pages). Which I saw earlier in the git and [nanomsg](https://github.com/nanomsg/nanomsg) projects (look under `doc`). The pipeline then goes like this:

```
.md --(asciidoc)--> .xml --(xmlto)--> .{1,2,...}
```

But `md2man` looks even better, awesome! I'm very glad we're going to get easily editable manpages now.

On a related note, I think either `man` or `doc` are not bad places for documentation like this.

noahfrederick commented on runtime/doc/manpages/nvim.1.md:
```diff
@@ -0,0 +1,261 @@
+
+## DESCRIPTION
+
+`nvim` is an extensible text editor based off of Vim.
```
I would change "based off of Vim" to "based on Vim." The former sounds overly colloquial to me.

splinterofchaos replied:
Would saying "forked off of" be too technical?

justinmk replied:
"based on" is best. Same wording is used in Vim, "based on Stevie".

Pyrohh replied:
:+1:

Pyrohh commented:
> But md2man looks even better, awesome! I'm very glad we're going to get easily editable manpages now.

Yes, indeed. Not having to break lines to use sane syntax (e.g., `\fB\-a\fR` inline vs. `.B \-a` on one line) makes it much easier to use semantic linefeeds, as well as see the general structure more easily.

> On a related note, I think either man or doc are not bad places for documentation like this.

Sounds good, I'll wait for some more advice just in case.

Pyrohh commented:
@jszakmeister and/or @fwalch could you take a look at e445de3c095bfc2c872121ace26074aa6d113fb8? The English man pages install fine for me, but I'm not really sure what I'm doing in some places.

I couldn't figure out how to generate the man pages using CMake, so I did it with this shell script and added the generated man pages to the PR (for now):
```
# (re)generate man pages
function man-regen {
	for file in *.md; do
		md2man-roff "${file}" >| "${file%.md}"
	done
}
```

fwalch commented:
I'd leave the changes from hard-coded to CMAKE_INSTALL_* paths for another PR, it's not directly related to manpage generation.

As an example how run an executable, see GenerateHelptags. Something like this should work for md2man-roff as well. But then everyone building neovim would need to have md2man installed, so.. Add it as a bundled dependency, like e.g. luarocks? Not sure about that. But it seems more appropriate than regenerating/checking in the man pages manually.

Pyrohh commented:
> I'd leave the changes from hard-coded to CMAKE_INSTALL_* paths for another PR, it's not directly related to manpage generation.

The man pages aren't installed atm... but I suppose I see your point since this also affects how everything else is installed. I'll do that

> As an example how run an executable, see GenerateHelptags. Something like this should work for md2man-roff as well.

I did, but I'm not sure how to pipe things and chop off suffixes in CMake. I can look into it. Is there any way to define what is to be iterated over in a `foreach` inline? i.e. `for file in *.md; do` instead of having to do this:
```
file(GLOB MANPAGES
  RELATIVE ${CMAKE_CURRENT_SOURCE_DIR}
  man/*.1)
```

> But then everyone building neovim would need to have md2man installed, so.. Add it as a bundled dependency, like e.g. luarocks? Not sure about that. I guess regenerating/checking in the man pages manually should be fine, they won't change that often.

I quite like the bundled dependency idea, but do we already have the infrastructure in place for ruby gems? I don't like having generated files in the source tree, as it requires them to be kept in sync (as you said). Not really sure about what to do here...

fwalch commented:
If you want to iterate over it, you should be able to do something like this:

```cmake
foreach(MANPAGE ${MANPAGES})
    message(STATUS "Processing ${MANPAGE}")
    # ...
endforeach()
```

I don't think you can get rid of the `file(GLOB ...)` call.

Adding Ruby and Rubygems to the built-in deps just for `md2man` seems like overkill... maybe we can get by without generating manpages from our own "developer" top-level Makefile (`make install`) by default, but enable it by default when CMake is used directly (i.e. what distribution packagers do).

Pyrohh commented:
> If you want to iterate over it, you should be able to do something like this: ...

Thanks, I'll look into that.

> Adding Ruby and Rubygems to the built-in deps just for md2man seems like overkill... maybe we can get by without generating manpages from our own "developer" top-level Makefile (make install) by default, but enable it by default when CMake is used directly (i.e. what distribution packagers do).

Yeah, I somewhat agree. Then again, such a distinction between the top-level Makefile and CMake seems non-obvious... this might be better discussed when relevant (as I'm going to move the most recent commit out of this PR).

I really wish we could use mandoc... :disappointed:

Pyrohh commented:
> I really wish we could use mandoc :disappointed: 

Well I guess I was wrong here. `mandoc` is a suite of tools revolving around the `mdoc` language. See [the mandoc website](http://mdocml.bsd.lv/) for more information. What's important is that `mandoc` isn't the only implementation of an `mdoc` parser: `groff` has an extensive `mdoc` implementation as well. Our targeted platforms (besides Windows) seem to all have recent enough versions for good mdoc support.

Anyways, I switched away from the markdown method because of the ruby dependency, but it's not like mdoc is actually difficult to write; if anything, I find it quite pleasant because of how much more control it gives than md2man. As I mentioned in the (edited) description, an excellent guide for the language is available here: http://manpages.bsd.lv/

Most of the advantages of mdoc compared to roff are documented in that guide, but the main difference is that roff consists primarily (exclusively?) of presentation macros, e.g., bold, italic, underline, while mdoc consists primarily (but not exclusively) of semantic macros, e.g., `.Fl` for flag, `.An` for author, `.Pa` for path, etc.

Commit:
Man pages: English: Convert to mdoc and rewrite for Nvim

Still needs some work.

- It would be great if Cmake could change the paths under FILES at
  compile time to be accurate, e.g. some systems will install to
  /usr/local/share, while others just /usr/share

- Add in nvim specific arguments

- Add Nvim authors under AUTHORS

Commit:
main.c: Make '-v' synonymous with '--version'

This is done for consistency with '-h' and '--help'.
Also update usage() to match man-pages.

Commit:
Man pages: Move to nvim namespace and out of `runtime/`

Also clean up man page references in `runtime/doc/Makefile` as it
doesn't deal with them anymore.

Commit:
$ git dump

Commit:
Use .Li macro (literal text) to prevent line breaks in the middle of 'options'

Commit:
fix incorrect macro usage

Pyrohh commented:
I audited the whole man page which ended up in a bunch of changes. This PR already has a bunch of comments which are no longer relevant, I'll open a new PR.
